src/riscv.c

```diff
@@ -0,0 +1,131 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include <assert.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "riscv_private.h"
+
+/* initialize the block map */
+static void block_map_init(struct block_map *map, const uint8_t bits)
+{
+    map->block_capacity = 1 << bits;
+    map->size = 0;
+    map->map = calloc(map->block_capacity, sizeof(struct block *));
+}
+
+riscv_user_t rv_userdata(struct riscv_t *rv)
+{
+    assert(rv);
+    return rv->userdata;
+}
+
+bool rv_set_pc(struct riscv_t *rv, riscv_word_t pc)
+{
+    assert(rv);
+#if RV32_HAS(EXT_C)
+    if (pc & 1)
+#else
+    if (pc & 3)
+#endif
+        return false;
+
+    rv->PC = pc;
+    return true;
+}
+
+riscv_word_t rv_get_pc(struct riscv_t *rv)
+{
+    assert(rv);
+    return rv->PC;
+}
+
+void rv_set_reg(struct riscv_t *rv, uint32_t reg, riscv_word_t in)
+{
+    assert(rv);
+    if (reg < RV_NUM_REGS && reg != rv_reg_zero)
+        rv->X[reg] = in;
+}
+
+riscv_word_t rv_get_reg(struct riscv_t *rv, uint32_t reg)
+{
+    assert(rv);
+    if (reg < RV_NUM_REGS)
+        return rv->X[reg];
+
+    return ~0U;
+}
+
+struct riscv_t *rv_create(const struct riscv_io_t *io, riscv_user_t userdata)
+{
+    assert(io);
+
+    struct riscv_t *rv = (struct riscv_t *) malloc(sizeof(struct riscv_t));
```

Comment:
For C programming, you don't have to cast the result (type: `void *`) of `malloc`. i.e., you can simply write as following:
```c
struct riscv_t *rv = malloc(sizeof(struct riscv_t));
```

Reply:
Furthermore, use `calloc` instead of combination of `malloc` and `memset`.